libsoftwarecontainer/src/netlink.cpp

```diff
@@ -0,0 +1,559 @@
+            fprintf(stderr, "Failed to set ip on link %i\n", ifinfo.ifi_index);
+            return ReturnCode::FAILURE;
+        } else {
+            return ReturnCode::SUCCESS;
+        }
+    }
+
+    // No link found, that's a failure.
+    return ReturnCode::FAILURE;
+}
+
+ReturnCode Netlink::linkDown(const int ifaceIndex)
+{
+    for (LinkInfo link : m_links) {
+        ifinfomsg ifinfo = link.first;
+        if(ifinfo.ifi_type == ARPHRD_LOOPBACK) {
+            continue; // This is the loopback device
+        }
+
+        netlink_request<ifinfomsg> down_msg = createMessage<ifinfomsg>(RTM_NEWLINK, 0);
+        down_msg.pay.ifi_family = AF_UNSPEC;
+        down_msg.pay.ifi_index = ifinfo.ifi_index;
+        down_msg.pay.ifi_flags = ~IFF_UP;
+        down_msg.pay.ifi_change = IFF_UP;
+        
+        if (isError(sendMessage(down_msg))) {
+            return ReturnCode::FAILURE;
+        } else {
+            return ReturnCode::SUCCESS;
+        }
+    }
+
+    return ReturnCode::FAILURE;
+}
+
+ReturnCode Netlink::findLink(const char *ifaceName, LinkInfo &linkInfo)
+{
+    if (isError(checkKernelDump())) {
+        return ReturnCode::FAILURE;
+    }
+
+    for (LinkInfo link : m_links) {
+        AttributeList attributes = link.second;
+        for (AttributeInfo attrinfo : attributes) {
+            rtattr attr = attrinfo.first;
+            void *data = attrinfo.second;
+
+            if (attr.rta_type == IFLA_IFNAME) {
+                char *ifname = (char *) data;
+                if (strcmp(ifname, ifaceName) == 0) {
+                    linkInfo = link;
+                    return ReturnCode::SUCCESS;
+                }
+            }
+        }
+    }
+
+    return ReturnCode::FAILURE;
+}
+
+ReturnCode Netlink::findAddresses(const unsigned int interfaceIndex, std::vector<AddressInfo> &result)
+{
+    if (isError(checkKernelDump())) {
+        return ReturnCode::FAILURE;
+    }
+
+    for (AddressInfo addressInfo : m_addresses) {
+        ifaddrmsg addrmsg = addressInfo.first;
+        if (addrmsg.ifa_index == interfaceIndex) {
+            result.push_back(addressInfo);
+        }
+    }
+
+    return ReturnCode::SUCCESS;
+
+}
+
+ReturnCode Netlink::hasAddress(const std::vector<AddressInfo> &haystack, const int addressFamily, const char *needle)
+{
+    for (AddressInfo addressInfo : haystack) {
+        AttributeList attributes = addressInfo.second;
+        for (AttributeInfo attrPair : attributes) {
+
+            // Skip if this attribute is not an address attribute
+            rtattr attr = attrPair.first;
+            if (attr.rta_type != IFA_ADDRESS && attr.rta_type != IFA_LOCAL) {
+                continue;
+            }
+
+            void *data = attrPair.second;
+            char out[INET6_ADDRSTRLEN];
+
+            // It is ok here for inet_ntop to fail (data could be bad)
+            if (inet_ntop(addressFamily, data, out, sizeof(out))) {
+                if (strcmp(out, needle) == 0) {
```

Comment:
if (a) if (b) ... == if (a && b)
Maybe it could be rewritten to increase readability ?